feat: move from websocket to ws and support native browser WebSocket API

Matsuuu opened the pull request:
## What kind of change does this PR introduce?

Make realtime-js runnable in the browser without requiring a build step.

Remove the old unmaintained `websocket` library and replace it with `ws` for node envs.

Make browsers use the native WebSocket interface.


## What is the current behavior?

The implementation of this library is not usable in the browser without a build step.

## What is the new behavior?

The implementation of this library is usable through the browser without the need to build your source code.

## Additional context

On my [trek through supabase codebase to make this project runnable in the browser](https://github.com/supabase/supabase-js/issues/671), this library was one of the two offenders of not working in the browser natively. 

Changed up the library in use to a more modern one, and made the browsers use a native interface for websockets for their communications.

Haven't done extensive manual testing but all of the tests passed green so I would expect this is quite a plug-in replacement for the old implementation.

Feel free to open a discussion on this. I really want to use Supabase but I also don't want to restrict myself by adding a toolchain to my dev flow.

w3b6x9 commented on src/RealtimeClient.ts:
```diff
@@ -132,7 +134,7 @@ export default class RealtimeClient {
       return
     }
 
-    this.conn = new this.transport(this._endPointURL(), [], null, this.headers)
```
@Matsuuu we'll still need to pass the headers for server-side: https://github.com/websockets/ws/issues/2103#issuecomment-1348594204

Matsuuu replied:
Fought some time with the type system, before I ended up scratching that as we were able to get by with this neat trick.

The WS api takes the options, where headers are put into, as the third arugment https://github.com/websockets/ws/blob/master/lib/websocket.js#L52

And the native websocket API doesn't have a third param, allowing us to do whatever we please with it https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/WebSocket

w3b6x9 replied:
@Matsuuu love it, thanks

w3b6x9 reviewed (changes requested):
@Matsuuu we've been meaning to do this  so thanks for tackling this! just one change requested.

w3b6x9 reviewed (approved):
@Matsuuu lgtm, thanks

w3b6x9 reviewed (changes requested):
@Matsuuu so the problem here is that the native WebSocket object API and ws API differ  so the changes in this PR will break when used on the client.

Matsuuu commented:
> @Matsuuu so the problem here is that the native WebSocket object API and ws API differ so the changes in this PR will break when used on the client.
> 
> discovered that while doing some manual testing on the server and client.

Kinda had a feeling that it wouldn't be this easy! :D 

I'll gladly help port the code to support both formats. Should we create a small repro program to test it against to find these cases out or update test suites to include these? 

I think a more clear separation of the types would help us find the differentiators.

How should we approach fixing this?

Matsuuu commented:
I'll try to find a time window to work on this. I think we need to change the API a little bit depending on what instance we're using. 

I'll try to make the typing around the websocket actually precise so it will report the errors and then work from there

Matsuuu commented:
Heya @w3b6x9 

I updated the typing around the components, fixed/extended the typings of other parts as needed. Ran tests and everything passed. Could you test out and report if you find some parts that are not functional?

Matsuuu commented:
Ran into a couple of errors while debugging with the demo, and fixed them in the latest commit. Feel free to give it a run!

w3b6x9 reviewed (approved):
@Matsuuu awesome work! i checked your latest commits against both server and client and both are working as expected. lgtm!

filipecabaco reviewed (approved):
🤩 this looks awesome!

mqp commented:
`websocket` doesn't build with Turbopack right now, so this is a big win for me.

Matsuuu commented:
After this gets merged, I think we've achieved the support for fully buildless setup for Supabase / supabase-js. (related https://github.com/supabase/supabase-js/issues/671).

Would be glad to write up on the subject and e.g. utilizing Supabase through import maps.

bmackio commented:
Big thank you @Matsuuu :)

@w3b6x9   Just curious if there is a RC and release date for this one? 

Per @filipecabaco comment above, would be great to put the node-gyp-build issue to rest or know whether I need to implement a workaround in the interim

robipop22 commented:
This commit is breaking builds on Vercel Edge routes because of the "ws" package.

w3b6x9 commented:
@robipop22 can you provide details? what errors are you getting?
